### PR TITLE
Upgrade tests to react-test-buffet

### LINKED ISSUE #23: Migrate unit tests to react-test-buffet
## Deliverables

- Reporter unit tests using react-test-buffet

#### Prerequisites
- [react-test-buffet](https://github.com/NiGhTTraX/react-test-buffet)

#### TODO
- [x] TBD

cc: @adela-istrate-hs

### PR DESCRIPTION
**Pull request by adela-istrate-hs**
Closes #23

### CODE REVIEW

**Comment by adela-istrate-hs** [comment]
@flavius-tirnacop-hs @valentin-radulescu-hs @andrei-picus-hs Can you review?

File: package.json [hunk 1]
```diff
@@ -26,21 +26,32 @@
...
     "classnames": "~2.2.5",
     "css-loader": "~0.19.0",
     "eslint": "~2.0.0",
     "eslint-plugin-react": "~4.1.0",
+    "jsdom": "~9.4.1",
     "lodash": "~3.10.1",
     "mocha": "~2.3.1",
     "mugshot": "^0.2.0",
     "object-assign": "~4.0.1",
+    "path": "~0.12.7",
```
Comment (andrei-picus-hs) [inline package.json]: `path` is a built in package.

File: package.json [hunk 2]
```diff
@@ -26,21 +26,32 @@
...
     "react": "~0.14.8",
+    "react-addons": "~0.9.1-deprecated",
```
Comment (andrei-picus-hs) [inline package.json]: This doesn't seem to be used. What were you using it for anyway?

File: package.json [hunk 3]
```diff
@@ -26,21 +26,32 @@
...
     "react-addons-test-utils": "~0.14.8",
+    "react-bootstrap": "~0.30.2",
```
Comment (andrei-picus-hs) [inline package.json]: This is already a production dep.

File: package.json [hunk 4]
```diff
@@ -26,21 +26,32 @@
...
+    "react-component-tree": "~0.4.0",
```
Comment (andrei-picus-hs) [inline package.json]: This should be a production dep as the reporter components depend on it.

File: tests/unit/runner.html [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        if (protocol !== 'file:') {
+          sourceMapSupport.install();
+        }
+      })();
+    </script>
+    <script src="../node_modules/mocha/mocha.js"></script>
+    <script>
+      mocha.setup('bdd');
+    </script>
+    <script src="../build/tests.js"></script>
```
Comment (andrei-picus-hs) [inline tests/unit/runner.html]: The setup for the HTML runner is not in this PR. I suggest taking the runner out as well and add it in a future issue.

File: tests/unit/helpers.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  _container = document.createElement('div');
+}
+
+
+/**
+ * Render a component into the DOM.
+ *
+ * @param {React class} Component
+ * @param {Object} fixture
+ * @param {DOM} container You should set this to this.container inside your
```
Comment (andrei-picus-hs) [inline tests/unit/helpers.js]: Remove this as it's no longer used. I made an [issue](https://github.com/NiGhTTraX/react-test-buffet/issues/5) to update upstream 😄

File: ui/components/failed.jsx [hunk 7]
```diff
@@ -54,31 +95,22 @@ class FailedTest extends React.Component {
...
-      const name = item.name;
+    FailedTest.VIEWS.forEach(function(item) {
       buttons.push(
-        <Button name={name}
-                key={name}
-                onClick={onViewChange}
-                className={classNames({active: name === currentView})}>
-          {name}
-        </Button>
+        _this.loadChild('selectViewButton', item)
```
Comment (andrei-picus-hs) [inline ui/components/failed.jsx]: Use fat arrows instead.

**Comment by flavius-tirnacop-hs** [comment]
image
wtf_thumb

File: ui/components/failed.jsx [hunk 8]
```diff
@@ -54,31 +95,22 @@ class FailedTest extends React.Component {
   }
 
   _renderSelectedView(paths) {
-    const Component =
-      _.find(FailedTest.VIEW_COMPONENTS, {name: this.state.view}).component;
 
     return <div>
-      <Jumbotron> <Component paths={paths} /> </Jumbotron>
+      <Jumbotron> {this.loadChild(this.state.view, paths)} </Jumbotron>
```
Comment (andrei-picus-hs) [inline ui/components/failed.jsx]: No me gusta. You're duplicating the setup between the children **and** you're moving the strategy details from outside to inside the component.

File: ui/components/header.jsx [hunk 9]
```diff
@@ -59,11 +59,13 @@ class Header extends React.Component {
...
-                 onClick={this.onFilterChange.bind(this, filter)}
                  className={classNames({active: this.props.filter === filter})}
-                 disabled={item.disabled}>
+                 disabled={item.disabled}
+                 ref={filter}
+                 onClick={this.onFilterChange.bind(this, filter)}>
         {item.name}
-        <Badge className={classNames(filter, item.badgeClass)}>
+        <Badge className={classNames(filter, item.badgeClass)}
+               ref={filter + 'Badge'}>
```
Comment (andrei-picus-hs) [inline ui/components/header.jsx]: Use template literals.

File: tests/unit/components/header/render.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import Header from '../../../../ui/components/header.jsx';
+import fixture from '../../../fixtures/components/header/header.js';
+import {render} from '../../helpers.js';
+
+const MENU_ITEMS = ['all', 'passes', 'failures'];
+
+describe('Header', function() {
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/header/render.js]: Please remove empty lines at the beginning of a statement block.

File: tests/unit/components/report/children.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    } = suitesParser(fixture.data);
+    let component, props;
+
+    beforeEach(function() {
+      stubMethod(Report, 'render', null);
+      component = render(Report, fixture);
+    });
+
+    describe('Header', function() {
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/children.js]: Please remove empty lines at the beginning of a statement block.

File: tests/unit/components/report/setup.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
+
+function _hasPassed(test) {
+  return test.state === 'passed';
+}
+
+module.exports = function(data) {
+  let numberOfPasses = 0,
+      numberOfFailures = 0,
+      duration = 0,
+      suites = data;
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/setup.js]: Just call the param `'suites'` directly.

File: tests/unit/components/report/setup.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  suites.forEach(function(suite) {
+    suite.passes = suite.tests.filter(function(test) {
+      return _hasPassed(test);
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/setup.js]: If all you're doing is passing the arguments received from the wrapping function then just use the inner function directly like `suite.tests.filter(_hasPassed)`.

File: tests/unit/components/report/setup.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    });
+    numberOfPasses += suite.passes.length;
+
+    suite.failures = suite.tests.filter(function(test) {
+      return !_hasPassed(test);
+    });
+    numberOfFailures += suite.failures.length;
+
+    suite.tests.forEach(function(test) {
+      duration += test.duration;
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/setup.js]:
```js
const getSuiteDuration = tests => tests.reduce(
  (acc, {duration}) => acc += duration, 0);
  
const totalDuration = suites.reduce((acc, {tests}) =>
  acc += getSuiteDuration(tests)
, 0);
```

You can calculate `numberOf{Failures,Passes}` in a similar fashion.

File: tests/unit/components/report/setup.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+      duration = 0,
+      suites = data;
+
+  suites.forEach(function(suite) {
+    suite.passes = suite.tests.filter(function(test) {
+      return _hasPassed(test);
+    });
+    numberOfPasses += suite.passes.length;
+
+    suite.failures = suite.tests.filter(function(test) {
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/setup.js]: You're mutating the `suite` object here. Do you really need this property? If so, clone the object first.

File: tests/unit/components/report/setup.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/report/setup.js]: Remove empty line.

File: tests/unit/components/failed/render.js [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+      function() {
+        expect(component.refs.errorButton.props.onClick)
+            .to.equal(component.onErrorMessageOpen);
+      });
+
+    it('should initially render the default view child', function() {
+      expect(loadChildStub).to.have.been.calledWith('default', fixture.paths);
+    });
+
+    FailedTest.VIEWS.slice(1, 4).forEach(function(item) {
```
Comment (andrei-picus-hs) [inline tests/unit/components/failed/render.js]: `slice(1)` will return all the elements except the first regardless of the length of the array.

File: tests/unit/components/details/passed/children.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
+import Details from '../../../../../ui/components/details.jsx';
+import fixture from '../../../../fixtures/components/details/details-passed.js';
+import {render, stubMethod, getChildProps} from '../../../helpers.js';
+
+describe('Details', function() {
+
+  describe('Children', function() {
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/details/passed/children.js]: Remove empty lines.

File: tests/unit/components/details/failed/render.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
+import Details from '../../../../../ui/components/details.jsx';
+import fixture from '../../../../fixtures/components/details/details-failed.js';
+import {render, stubMethod} from '../../../helpers.js';
+
+describe('Details', function() {
+
+  describe('Render', function() {
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/details/failed/render.js]: Remove empty lines.

File: tests/unit/components/details/failed/children.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+      it('should pass the diff path', function() {
+        expect(childPaths.diff).to.equal(expectedPaths.diff);
+      });
+
+      it('should pass an error object', function() {
+        expect(props.error).to.equal(expectedError);
+      });
+    });
+  });
+
```
Comment (andrei-picus-hs) [inline tests/unit/components/details/failed/children.js]: Remove empty line.

File: tests/unit/.eslintrc [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    'mocha': true
+  },
+
+  'globals': {
+    'expect': false,
+    'sandbox': false,
+    'document': false,
+    'window': false,
+    'React': false,
+    'ReactDOM': false,
```
Comment (andrei-picus-hs) [inline tests/unit/.eslintrc]: `React`, `ReactDOM` are not globals. And `TestUtils` is not even used (at least not in this PR).

File: tests/fixtures/components/views/paths.js [hunk 22]
```diff
@@ -0,0 +1,5 @@
+import fixture from '../failed/failed.js';
+
+module.exports = {
```
Comment (andrei-picus-hs) [inline tests/fixtures/components/views/paths.js]: If we're going to use `import` statements then let's use `export` statements as well.

File: tests/unit/components/views/swipe/render.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+    });
+
+    it('should render an input range with the default value ' +
+      `${SwipeView.DEFAULT_VALUE}`, function() {
+      expect(inputRangeProps.value).to.equal(`${SwipeView.DEFAULT_VALUE}`);
+    });
+
+    it('should have attached on the inputRange the onValueChange cb',
+      function() {
+        expect(component.refs.inputRange.props.onChange)
```
Comment (andrei-picus-hs) [inline tests/unit/components/views/swipe/render.js]:
```
Warning: ReactDOMComponent: Do not access .props of a DOM node;
instead, recreate the props as `render` did originally or read the DOM
properties/attributes directly from this node (e.g., this.refs.box.className).
This DOM node was rendered by `SwipeView`.
```

Comment (andrei-picus-hs) [inline tests/unit/components/views/swipe/render.js]: [NiGhTTraX/react-test-buffet/tests/setup/react-warning.js](https://github.com/NiGhTTraX/react-test-buffet/blob/750c8088520153d68c9e67a6e1eb26cb0f818717/tests/setup/react-warnings.js) comes in handy for this kind of thing.

**Comment by andrei-picus-hs** [comment]
Finished review.

File: tests/unit/components/views/swipe/render.js (hunk 23 above)
Comment (valentin-radulescu-hs) [inline tests/unit/components/views/swipe/render.js]: Let's get around this warning by setting a value on the input and checking that the state has been updated accordingly:

```js
component.refs.inputRange.value = randomValue;
TestUtils.Simulate.change(component.refs.inputRange);

expect(component.state.value).to.equal(randomValue);
```

File: package.json (hunk 2 above)
Comment (adela-istrate-hs) [inline package.json]: `addons.classSet` was used in the `Test.jsx` component. Now I'm using `classNames` and I forgot to remove the addons package 😄  .

File: tests/unit/.eslintrc (hunk 21 above)
Comment (adela-istrate-hs) [inline tests/unit/.eslintrc]: I would like to import `React` and `ReactDOM` only once, not in every file. How can I make the linter not scream about it?

Comment (andrei-picus-hs) [inline tests/unit/.eslintrc]: Why hide dependencies? If you need React in your module, then make it explicit. Implicit dependencies are never a good thing. Think about what happens when you move your module to an environment that doesn't expose React as a global. It's this kind of unexpected things that make maintaining code harder.

File: package.json [hunk 24]
```diff
@@ -26,21 +26,27 @@
...
     "mocha": "~2.3.1",
     "mugshot": "^0.2.0",
     "object-assign": "~4.0.1",
     "react": "~0.14.8",
-    "react-addons-test-utils": "~0.14.8",
     "react-dom": "~0.14.8",
     "react-image-diff": "~0.1.0",
+    "sinon": "github:ubervu/Sinon.JS",
+    "sinon-chai": "~2.8.0",
     "skin-deep": "~0.6.0",
```
Comment (valentin-radulescu-hs) [inline package.json]: You can remove the `skin-deep` package as it's not being used anymore.

File: ui/components/results.jsx [hunk 25]
```diff
@@ -1,11 +1,27 @@
...
+        };
+      }
+    };
+  }
 
-class Results extends React.Component {
   render() {
-    const filter = this.props.filter;
+    const filter = this.props.filter,
+          _this = this;
```
Comment (valentin-radulescu-hs) [inline ui/components/results.jsx]: You can avoid having to hoist `this` here and in other tests by using fat arrows.

**Comment by valentin-radulescu-hs** [comment]
Finished review.

File: ui/components/failed.jsx [hunk 26]
```diff
@@ -95,9 +77,11 @@ class FailedTest extends Component {
   }
 
   _renderSelectedView(paths) {
+    const component =
+      _.find(FailedTest.VIEWS, {name: this.state.view}).component;
 
     return <div>
-      <Jumbotron> {this.loadChild(this.state.view, paths)} </Jumbotron>
+      <Jumbotron> {this.loadChild('view', component, paths)} </Jumbotron>
```
Comment (adela-istrate-hs) [inline ui/components/failed.jsx]: @andrei-picus-hs Te gusta?

Comment (andrei-picus-hs) [inline ui/components/failed.jsx]: Si sombrero.

**Pull request merged**